pkg/server/instancecleaner/cleaner.go

```diff
@@ -0,0 +1,91 @@
+package instancecleaner
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gocardless/draupnir/pkg/exec"
+	"github.com/gocardless/draupnir/pkg/models"
+	"github.com/gocardless/draupnir/pkg/store"
+	"github.com/prometheus/common/log"
+
+	google "google.golang.org/api/oauth2/v1"
+)
+
+type InstanceCleaner struct {
+	done          chan bool
+	logger        log.Logger
+	instanceStore store.InstanceStore
+	executor      exec.Executor
+}
+
+func NewInstanceCleaner(logger log.Logger, instanceStore store.InstanceStore, executor exec.Executor) *InstanceCleaner {
+	done := make(chan bool)
+	return &InstanceCleaner{
+		done:          done,
+		logger:        logger.With("component", "cleaner"),
+		instanceStore: instanceStore,
+		executor:      executor,
+	}
+}
+
+func (ic *InstanceCleaner) Start(cctx context.Context, interval time.Duration) error {
+	for {
+		select {
+		case <-time.After(interval):
+			ic.logger.Info("cleaning old instances with invalid tokens")
+			instances, err := ic.instanceStore.List()
+			if err != nil {
+				ic.logger.Info("skipping cleaning old instances - could not get instances list")
+			} else {
+				for _, instance := range instances {
+					if instance.AccessToken != "" {
+						valid, err := validateAccessToken(instance.AccessToken)
+						if err != nil {
+							ic.logger.With("instance", instance.ID).Error("failed to validate token: %s", err.Error())
+						} else if !valid {
+							ic.logger.With("instance", instance.ID).Info("token for instance invalid - destroying instance")
+							err = ic.destroyInstance(instance)
+							if err != nil {
+								ic.logger.With("instance", instance.ID).Error("failed to destroy instance: %s", err.Error())
+							}
+						}
+					}
+				}
+			}
+		case <-cctx.Done():
+			ic.done <- true
+			return nil
+		}
+	}
+}
+
+func (ic *InstanceCleaner) Stop() {
```

Comment:
I don't think you need this? I think you can get away with a single method `Start(context.Context)` which loops until the context expires, then terminates.

Then the rungroup can just cancel the context and wait for the primary `Start()` function to die.

Reply:
Yep, with changing around how the context and run group is used this isn't needed.